### PR TITLE
Enable/disable all interfaces from a switch

### LINKED ISSUE #79: Enable/disable all interfaces from a switch
Topology NApp should have a REST service to enable all interfaces from switch. A switch can have 128 interfaces, and it is cumbersome to enable/disable each one.
The current service enable/disable just one interface:
https://github.com/kytos/topology/blob/master/main.py#L153

### PR DESCRIPTION
**Pull request by cmagnobarbosa**
Now user can enable/disable all interfaces from a switch. To do this, user can make request by REST endpoints:

- Enabling all interfaces on a switch:
```
POST http://localhost:8181/api/kytos/topology/v3/interfaces/switch/{dpid}/enable
```
- Disabling all interfaces on a switch:
```
POST http://localhost:8181/api/kytos/topology/v3/interfaces/switch/{dpid}/disable
```
Obs: `dpid `is the switch id.

Related: #79

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -71,6 +71,14 @@ def _get_link_from_interface(self, interface):
                 return link
         return None
 
+    @staticmethod
+    def _get_data(req):
+        """Get request data."""
+        data = req.get_json()  # Valid format { "interfaces": [...] }
```
Comment (gleybersonandrade) [inline main.py]: data = req.get_json() or {}

**Review by ajoaoff — COMMENTED** [review]
This PR requests the user to send in the POST payload the list of interfaces, but if the user wants to delete all interfaces from a switch, it needed only to send the switch DPID and the application should get the list of interfaces of that switch.

File: main.py [hunk 2]
```diff
@@ -45,6 +45,15 @@ def _get_link_or_create(self, endpoint_a, endpoint_b):
...
 
+    def _get_interfaces(self, dpid):
+        """Return a list all interfaces from a switch."""
+        try:
+            switch = self.controller.switches[dpid]
+        except KeyError as exc:
+            raise KeyError(f"Switch not found :{exc}")
+        return [iface.as_dict().get('id') for iface in
+                switch.interfaces.values()]
+
```
Comment (ajoaoff) [inline main.py]: This method is not necessary.

File: main.py [hunk 3]
```diff
@@ -155,60 +164,58 @@ def get_interfaces(self):
...
 
-    @rest('v3/interfaces/enable', methods=['POST'])
+    @rest('v3/interfaces/switch/<dpid>/enable', methods=['POST'])
     @rest('v3/interfaces/<interface_enable_id>/enable', methods=['POST'])
-    def enable_interface(self, interface_enable_id=None):
+    def enable_interface(self, interface_enable_id=None, dpid=None):
         """Administratively enable a list of interfaces in the topology."""
         error_list = []  # List of interfaces that were not activated.
         interface_ids = []
         msg_error = "Some interfaces couldn't be found and activated: "
```
Comment (ajoaoff) [inline main.py]:
```suggestion
        error_list = []  # List of interfaces that were not activated.
        msg_error = "Some interfaces couldn't be found and activated: "
```

You don't need to keep a list of interfaces.

File: main.py [hunk 4]
```diff
@@ -155,60 +164,58 @@ def get_interfaces(self):
...
+            if dpid:
+                interface_ids = self._get_interfaces(dpid)
+            else:
+                interface_ids.append(interface_enable_id)
+            for interface_id in interface_ids:
+                switch_id = ":".join(interface_id.split(":")[:-1])
+                interface_number = int(interface_id.split(":")[-1])
+            switch = self.controller.switches[switch_id]
+        except KeyError as exc:
+            return jsonify(f"Switch not found: {exc}"), 404
```
Comment (ajoaoff) [inline main.py]: There is too much code inside the try ... except. You should here just get the switch instance.

```suggestion
        if dpid is None:
            dpid = ":".join(interface_enable_id.split(":")[:-1])
        try:
            switch = self.controller.switches[dpid]
        except KeyError as exc:
            return jsonify(f"Switch not found: {exc}"), 404
```

File: main.py [hunk 5]
```diff
@@ -155,60 +164,58 @@ def get_interfaces(self):
...
 
-            try:
-                switch.interfaces[interface_number].enable()
-            except KeyError as exc:
-                error_list.append(f"Switch {switch_id} Interface {exc}")
+        try:
+            switch.interfaces[interface_number].enable()
+        except KeyError as exc:
+            error_list.append(f"Switch {switch_id} Interface {exc}")
```
Comment (ajoaoff) [inline main.py]: Here you need to traverse the whole list of interfaces to enable them all, or just one interface.

```suggestion
        if interface_enable_id:
            interface_number = int(interface_enable_id.split(":")[-1])
            try:
                switch.interfaces[interface_number].enable()
            except KeyError as exc:
                error_list.append(f"Switch {switch_id} Interface {exc}")
        else:
            for interface in switch.interfaces.values():
                interface.enable()
```

File: main.py [hunk 6]
```diff
@@ -155,60 +164,58 @@ def get_interfaces(self):
...
+            return jsonify(f"Switch not found: {exc}"), 404
 
-            try:
-                switch.interfaces[interface_number].disable()
-            except KeyError as exc:
-                error_list.append(f"Switch {switch_id} Interface {exc}")
+        try:
+            switch.interfaces[interface_number].disable()
+        except KeyError as exc:
+            error_list.append(f"Switch {switch_id} Interface {exc}")
```
Comment (ajoaoff) [inline main.py]: Same changes as in `enable_interface`.

**Review by ajoaoff — CHANGES_REQUESTED** [review]

File: main.py (hunk 6 above)
Comment (cmagnobarbosa) [reply main.py]: All suggested changes have been implemented.

**Review by ajoaoff — CHANGES_REQUESTED** [review]
There is no test of the new methods. Please, create unit tests for them.

**Review by ajoaoff — CHANGES_REQUESTED** [review]
The result in the tests should also be checked, i. e., check if the affected interfaces were really enabled/disabled. To do this, you could mock the calls to `enable()` and `disable()` and check how many times they were called.

**Comment by cmagnobarbosa** [comment]
> The result in the tests should also be checked, i. e., check if the affected interfaces were really enabled/disabled. To do this, you could mock the calls to `enable()` and `disable()` and check how many times they were called.

These tests now check whether an interface has been enabled or disabled correctly, counting the number of times the` enable ()` and `disable () `methods were called.

Commit: https://github.com/kytos/topology/pull/80/commits/f65c11d

Thank you a lot for the suggestion @ajoaoff, this PR has improved a lot.

File: main.py (hunk 1 above)
Comment (cmagnobarbosa) [reply main.py]: Thanks for suggestion @gleybersonandrade.

**Review by ajoaoff — APPROVED** [review]

**Pull request merged**